### ISSUE TITLE
bug: `POST /api/build` rebuilds all three models (TF-IDF, SVD, VADER) from scratch on every call — rebuilding takes significant time on large datasets and blocks the API server during model training

### ISSUE THREAD

**Issue by prince-pokharna** [leonagoel/hybrid-recommender#1766]
## 🐛 Problem Statement

The `/api/build` endpoint trains all three ML models (TF-IDF content vectorizer, SVD collaborative model, VADER sentiment aggregation) on every call. Based on `backend/main.py` and the model files (`content_model.py`, `collaborative_model.py`, `nlp_engine.py`), this training is synchronous — it blocks the FastAPI event loop during the entire model fitting process.

On a dataset of 25,000+ products (as advertised in the README: *"25,000+ products"*), this means:
1. The API is **completely unresponsive** to all other requests during model training
2. A user who accidentally clicks "Build Models" twice fires two concurrent training jobs that mutate the same in-memory model objects — causing a race condition
3. There is no progress feedback — the UI shows nothing until training completes

## Proposed Fix

**1. Run model training in a background thread using FastAPI's `run_in_executor`:**

```python
# backend/main.py
import asyncio
from concurrent.futures import ThreadPoolExecutor

executor = ThreadPoolExecutor(max_workers=1)  # Single worker prevents concurrent builds

build_status = {"is_building": False, "progress": 0, "error": None}

@app.post("/api/build")
async def build_models():
    if build_status["is_building"]:
        return JSONResponse({"error": "Build already in progress"}, status_code=409)

    build_status["is_building"] = True
    build_status["progress"] = 0
    build_status["error"] = None

    loop = asyncio.get_event_loop()
    asyncio.create_task(
        loop.run_in_executor(executor, _run_build_sync)
    )
    return {"message": "Build started in background", "poll": "/api/build/status"}

def _run_build_sync():
    try:
        build_status["progress"] = 10
        content_model.fit(products)
        build_status["progress"] = 40
        collab_model.fit(interactions)
        build_status["progress"] = 70
        nlp_engine.compute_sentiments(reviews)
        build_status["progress"] = 100
    except Exception as e:
        build_status["error"] = str(e)
    finally:
        build_status["is_building"] = False

@app.get("/api/build/status")
async def get_build_status():
    return build_status
```

**2. Add a `/api/build/status` polling endpoint** so the frontend can show a real progress bar.

**3. Update `frontend/app.js`** to poll `/api/build/status` every 2 seconds after clicking "Build Models" and display progress to the user.

## Files to Modify

| File | Change |
|---|---|
| `backend/main.py` | Move build to background thread, add status endpoint |
| `frontend/app.js` | Poll status endpoint, show progress bar during build |

**Suggested labels:** `bug`, `performance`, `backend`, `level: intermediate`

I would like to work on this. Could you please assign it to me?

**Comment by prince-pokharna** [comment]
Hey @leonagoel ,
Please assign it to me under GSSOC 2026.
/assign /claim

**Comment by karrisanthoshigayatri** [comment]
Hi @leonagoel,

I would like to work on this issue under GSSoC 2026. Could you please assign it to me?

**Approach:**

* Move model building to a background thread to keep the API responsive.
* Add a `/api/build/status` endpoint to track build progress and prevent concurrent builds.
* Update the frontend to poll the status endpoint and display a progress bar during model training.

Thank you!

**Comment by mahek-builds** [comment]
Hey @leonagoel Please assign me under GSSoC 2026. I want to work on this issue.
